### PR TITLE
Feature/cache pd status

### LINKED ISSUE #942: TiUP cluster display <cluster-name> 等待时间过长
## General Question

Before asking a question, make sure you have:

```
tiup --version
v1.2.3 tiup
Go Version: go1.13
Git Branch: release-1.2
GitHash: df7e28a
```

TiUP cluster display <cluster-name> 执行后，等待时间 2 分钟 5 秒

```
  $ tiup cluster display tc1
Cluster type:    tidb
Cluster name:    tc1
Cluster version: v4.0.8
SSH type:        builtin

WARN: get location labels from pd failed: no endpoint available, the last err is: Get http://172.16.4.90:22379/pd/api/v1/config/replicate: dial tcp 172.16.4.90:22379: connect: no route to host

real    2m5.929s
user    0m0.506s
sys     0m0.126s
```

cluster display 操作时，每个组件需要与 PD 通信并确认状态。
集群一般部署三个 PD 节点，每个组件会依次向 PD 发送请求获取信息，即 `组件数量 * PD 数量 * 10s` 后显示集群信息。

// 集群网络或组件异常时，为避免等待时间过长。是否优化为  `TiUP cluster display meta <cluster-name> `，不输出 Status 列，快速查看集群拓扑信息。

```
ID                  Role          Host          Ports        OS/Arch        Data Dir                         Deploy Dir
--                  ----          ----          -----        -------        --------                         ----------
172.16.5.103:3000   grafana       172.16.5.103  3000         linux/x86_64   -                                /opt/tc2/deploy/grafana-3000
172.16.4.90:22379   pd            172.16.4.90   22379/22380  linux/x86_64   /data1/tc2/data/pd-22379         /data1/tc2/deploy/pd-22379
```

### PR DESCRIPTION
**Pull request by 9547**
### What problem does this PR solve? 

- fix #942 
- implement #961 

### Check List 

Tests 

The old one need ~2minutes

```bash
root@control:/tiup-cluster# time bin/tiup-cluster display c1
Cluster type:    tidb
Cluster name:    c1
Cluster version: v4.0.4
SSH type:        builtin

ID                  Role            Host          Ports                            OS/Arch       Status  Data Dir                                                 Deploy Dir
--                  ----            ----          -----                            -------       ------  --------                                                 ----------
172.19.0.101:9093   alertmanager    172.19.0.101  9093/9094                        linux/x86_64  Up      /home/tidb/deploy/alertmanager-9093/data                 /home/tidb/deploy/alertmanager-9093
172.19.0.103:8300   cdc             172.19.0.103  8300                             linux/x86_64  Down    -                                                        /home/tidb/deploy/cdc-8300
172.19.0.104:8300   cdc             172.19.0.104  8300                             linux/x86_64  Down    -                                                        /home/tidb/deploy/cdc-8300
172.19.0.105:8300   cdc             172.19.0.105  8300                             linux/x86_64  Down    -                                                        /home/tidb/deploy/cdc-8300
172.19.0.101:8249   drainer         172.19.0.101  8249                             linux/x86_64  Up      /home/tidb/data/drainer-8249/data                        /home/tidb/deploy/drainer-8249
172.19.0.101:3000   grafana         172.19.0.101  3000                             linux/x86_64  Up      -                                                        /home/tidb/deploy/grafana-3000
172.19.0.103:2379   pd              172.19.0.103  2379/2380                        linux/x86_64  Down    /home/tidb/deploy/pd-2379/data                           /home/tidb/deploy/pd-2379
172.19.0.104:2379   pd              172.19.0.104  2379/2380                        linux/x86_64  Down    /home/tidb/deploy/pd-2379/data                           /home/tidb/deploy/pd-2379
172.19.0.105:2379   pd              172.19.0.105  2379/2380                        linux/x86_64  Down    /home/tidb/deploy/pd-2379/data                           /home/tidb/deploy/pd-2379
172.19.0.101:9090   prometheus      172.19.0.101  9090                             linux/x86_64  Up      /home/tidb/deploy/prometheus-9090/data                   /home/tidb/deploy/prometheus-9090
172.19.0.102:9090   prometheus      172.19.0.102  9090                             linux/x86_64  Up      /home/tidb/deploy/prometheus-9090/data                   /home/tidb/deploy/prometheus-9090
172.19.0.103:8250   pump            172.19.0.103  8250                             linux/x86_64  Up      /home/tidb/deploy/pump-8250/data                         /home/tidb/deploy/pump-8250
172.19.0.104:8250   pump            172.19.0.104  8250                             linux/x86_64  Up      /home/tidb/deploy/pump-8250/data                         /home/tidb/deploy/pump-8250
172.19.0.105:8250   pump            172.19.0.105  8250                             linux/x86_64  Up      /home/tidb/deploy/pump-8250/data                         /home/tidb/deploy/pump-8250
172.19.0.101:4000   tidb            172.19.0.101  4000/10080                       linux/x86_64  Up      -                                                        /home/tidb/deploy/tidb-4000
172.19.0.102:4000   tidb            172.19.0.102  4000/10080                       linux/x86_64  Up      -                                                        /home/tidb/deploy/tidb-4000
172.19.0.103:9000   tiflash         172.19.0.103  9000/8123/3930/20170/20292/8234  linux/x86_64  Down    /home/tidb/deploy/tiflash-9000/data1,/data/tiflash-data  /home/tidb/deploy/tiflash-9000
172.19.0.101:20160  tikv            172.19.0.101  20160/20180                      linux/x86_64  Down    /home/tidb/deploy/tikv-20160/data                        /home/tidb/deploy/tikv-20160
172.19.0.103:20160  tikv            172.19.0.103  20160/20180                      linux/x86_64  Down    /home/tidb/my_kv_data                                    /home/tidb/deploy/tikv-20160
172.19.0.104:20160  tikv            172.19.0.104  20160/20180                      linux/x86_64  Down    /home/tidb/deploy/tikv-20160/data                        /home/tidb/deploy/tikv-20160
172.19.0.105:20160  tikv            172.19.0.105  20160/20180                      linux/x86_64  Down    /home/tidb/deploy/tikv-20160/data                        /home/tidb/deploy/tikv-20160
172.19.0.103:7077   tispark-master  172.19.0.103  7077/8080                        linux/x86_64  Up      -                                                        /home/tidb/deploy/tispark-master-7077
172.19.0.104:7078   tispark-worker  172.19.0.104  7078/8081                        linux/x86_64  Up      -                                                        /home/tidb/deploy/tispark-worker-7078
Total nodes: 23

real    1m56.029s
user    0m0.287s
sys     0m0.133s
```

**After cached, the used time is limited in 3*10s**

```bash
root@control:/tiup-cluster# time bin/tiup-cluster display c1
Cluster type:       tidb
Cluster name:       c1
Cluster version:    v4.0.4
SSH type:           builtin
ID                  Role            Host          Ports                            OS/Arch       Status  Data Dir                                                 Deploy Dir
--                  ----            ----          -----                            -------       ------  --------                                                 ----------
172.19.0.101:9093   alertmanager    172.19.0.101  9093/9094                        linux/x86_64  Up      /home/tidb/deploy/alertmanager-9093/data                 /home/tidb/deploy/alertmanager-9093
172.19.0.103:8300   cdc             172.19.0.103  8300                             linux/x86_64  Down    -                                                        /home/tidb/deploy/cdc-8300
172.19.0.104:8300   cdc             172.19.0.104  8300                             linux/x86_64  Down    -                                                        /home/tidb/deploy/cdc-8300
172.19.0.105:8300   cdc             172.19.0.105  8300                             linux/x86_64  Down    -                                                        /home/tidb/deploy/cdc-8300
172.19.0.101:8249   drainer         172.19.0.101  8249                             linux/x86_64  Up      /home/tidb/data/drainer-8249/data                        /home/tidb/deploy/drainer-8249
172.19.0.101:3000   grafana         172.19.0.101  3000                             linux/x86_64  Up      -                                                        /home/tidb/deploy/grafana-3000
172.19.0.103:2379   pd              172.19.0.103  2379/2380                        linux/x86_64  Down    /home/tidb/deploy/pd-2379/data                           /home/tidb/deploy/pd-2379
172.19.0.104:2379   pd              172.19.0.104  2379/2380                        linux/x86_64  Down    /home/tidb/deploy/pd-2379/data                           /home/tidb/deploy/pd-2379
172.19.0.105:2379   pd              172.19.0.105  2379/2380                        linux/x86_64  Down    /home/tidb/deploy/pd-2379/data                           /home/tidb/deploy/pd-2379
172.19.0.101:9090   prometheus      172.19.0.101  9090                             linux/x86_64  Up      /home/tidb/deploy/prometheus-9090/data                   /home/tidb/deploy/prometheus-9090
172.19.0.102:9090   prometheus      172.19.0.102  9090                             linux/x86_64  Up      /home/tidb/deploy/prometheus-9090/data                   /home/tidb/deploy/prometheus-9090
172.19.0.103:8250   pump            172.19.0.103  8250                             linux/x86_64  Up      /home/tidb/deploy/pump-8250/data                         /home/tidb/deploy/pump-8250
172.19.0.104:8250   pump            172.19.0.104  8250                             linux/x86_64  Up      /home/tidb/deploy/pump-8250/data                         /home/tidb/deploy/pump-8250
172.19.0.105:8250   pump            172.19.0.105  8250                             linux/x86_64  Up      /home/tidb/deploy/pump-8250/data                         /home/tidb/deploy/pump-8250
172.19.0.101:4000   tidb            172.19.0.101  4000/10080                       linux/x86_64  Up      -                                                        /home/tidb/deploy/tidb-4000
172.19.0.102:4000   tidb            172.19.0.102  4000/10080                       linux/x86_64  Up      -                                                        /home/tidb/deploy/tidb-4000
172.19.0.103:9000   tiflash         172.19.0.103  9000/8123/3930/20170/20292/8234  linux/x86_64  N/A     /home/tidb/deploy/tiflash-9000/data1,/data/tiflash-data  /home/tidb/deploy/tiflash-9000
172.19.0.101:20160  tikv            172.19.0.101  20160/20180                      linux/x86_64  N/A     /home/tidb/deploy/tikv-20160/data                        /home/tidb/deploy/tikv-20160
172.19.0.103:20160  tikv            172.19.0.103  20160/20180                      linux/x86_64  N/A     /home/tidb/my_kv_data                                    /home/tidb/deploy/tikv-20160
172.19.0.104:20160  tikv            172.19.0.104  20160/20180                      linux/x86_64  N/A     /home/tidb/deploy/tikv-20160/data                        /home/tidb/deploy/tikv-20160
172.19.0.105:20160  tikv            172.19.0.105  20160/20180                      linux/x86_64  N/A     /home/tidb/deploy/tikv-20160/data                        /home/tidb/deploy/tikv-20160
172.19.0.103:7077   tispark-master  172.19.0.103  7077/8080                        linux/x86_64  Up      -                                                        /home/tidb/deploy/tispark-master-7077
172.19.0.104:7078   tispark-worker  172.19.0.104  7078/8081                        linux/x86_64  Up      -                                                        /home/tidb/deploy/tispark-worker-7078
Total nodes: 23

real    0m26.010s
user    0m0.209s
sys     0m0.122s
```

#### One more thing

for the dm's cluster, I've highlighted the dm's status code, similar to tidb-cluster's, like below:

image

And if dm-masters were down, the dm-workers' status previous as below:

image

after this or, changed to: (**Down** -> **N/A**, means worker's status depends on master, but all masters were down)

image

#### Dashboard url

display with dashboard-url appended 

```bash
root@control:/tiup-cluster# tiup-cluster display xxx
Cluster type:       tidb
Cluster name:       xxx
Cluster version:    v4.0.4
SSH type:           builtin
Dashboard URL:      http://n5:2379/dashboard
ID        Role            Host  Ports                            OS/Arch       Status  Data Dir                                  Deploy Dir
...
```
Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

Release notes:

```release-note
* enhancement: speed up the `{tiup|dm} cluster display;
* feature: Add dashboard URL in `tiup cluster display`;
* feature: highlight dm's status in `dm cluster display`;
```

### CODE REVIEW

**Comment by breeswish** [comment]
Hi, could you update your PR description about the Dashboard URL change?

**Comment by 9547** [comment]
> Hi, could you update your PR description about the Dashboard URL change?

Sorry, and I'll add it in the `Release notes`, it's anything else I'm missing?

**Comment by breeswish** [comment]
> > Hi, could you update your PR description about the Dashboard URL change?
> 
> Sorry, and I'll add it in the `Release notes`, it's anything else I'm missing?

Actually I only want to know what it looks like in the terminal output for this Dashbaord URL change :)

**Comment by 9547** [comment]
@breeswish PTAL 

```bash
root@control:/tiup-cluster# tiup-cluster display xxx
Cluster type:       tidb
Cluster name:       xxx
Cluster version:    v4.0.4
SSH type:           builtin
Dashboard URL:      http://n5:2379/dashboard
ID        Role            Host  Ports                            OS/Arch       Status  Data Dir                                  Deploy Dir
...
```

**Comment by lucklove** [comment]
Maybe we can remove the `--dashboard` flag if we merged it into normal display? @breeswish

**Comment by 9547** [comment]
> Maybe we can remove the `--dashboard` flag if we merged it into normal display? @breeswish

LGTM

File: pkg/cluster/spec/pd.go [hunk 1]
```diff
@@ -52,30 +51,19 @@ type PDSpec struct {
...
 
 // Status queries current status of the instance
-func (s PDSpec) Status(tlsCfg *tls.Config, pdList ...string) string {
-	curAddr := fmt.Sprintf("%s:%d", s.Host, s.ClientPort)
-	curPdAPI := api.NewPDClient([]string{curAddr}, statusQueryTimeout, tlsCfg)
-	allPdAPI := api.NewPDClient(pdList, statusQueryTimeout, tlsCfg)
+func (s PDSpec) Status(tlsCfg *tls.Config, _ ...string) string {
+	addr := fmt.Sprintf("%s:%d", s.Host, s.ClientPort)
+	pc := api.NewPDClient([]string{addr}, statusQueryTimeout, tlsCfg)
 	suffix := ""
```
Comment (lucklove) [inline pkg/cluster/spec/pd.go:57]: The suffix will always be empty since you delete the `find dashbaord` stage. So I think we can delete it this time

Comment (9547) [reply pkg/cluster/spec/pd.go]: PTAL

**Pull request merged**